File: steering/src/steerforce.py
#!/usr/bin/env python3

import rospy
from dataclasses import dataclass

from evdev import InputDevice, ecodes, ff
from steering.msg import SteerCtl

@dataclass
class DEVICE:
    NAME = "/dev/input/event14"
    LOOP_RATE = 0.1    
    TOLERANCE = 0.01
    MAX_TORQUE = 1.0
    MIN_TORQUE = 0.2
    
# class SteerCtl:
#     position = 0
#     springrate = 0

# Returns a value between lLim and uLim
def limit(value, lLim, uLim):
    return min(max(lLim, value), uLim)

class SteeringFeedback:
    def __init__(self):
        global DEVICE
        
        self.ns = rospy.get_namespace()        
        rospy.init_node("steering_feedback")

        DEVICE.NAME = rospy.get_param("~device_name", DEVICE.NAME)
        DEVICE.LOOP_RATE = rospy.get_param("~rate", DEVICE.LOOP_RATE)        
        DEVICE.TOLERANCE = rospy.get_param("~tolerance", DEVICE.TOLERANCE)

        DEVICE.MIN_TORQUE = rospy.get_param("~min_torque", DEVICE.MIN_TORQUE)
        DEVICE.MAX_TORQUE = rospy.get_param("~max_torque", DEVICE.MAX_TORQUE)

        self.p_des = 0 # Desired Positon 
        self.p_err = 0               

        self.k_des = 1 # Spring rate
        self.t_corr = 0 # Correcting Torque
        
        self.device = InputDevice(DEVICE.NAME)        
        self.axis_code = ecodes.ABS_X

        axis_info = self.device.absinfo(self.axis_code)
        
        self.axis_min = axis_info.min
        self.axis_max = axis_info.max

        self.device.write(ecodes.EV_FF, ecodes.FF_AUTOCENTER, 0)

        # Create and upload force feedback effect
        self.effect = ff.Effect(
            ecodes.FF_CONSTANT,
            -1,
            ff.Trigger(0, 0),
            ff.Replay(0xffff, 0),
            ff.Constant(level=0, envelope=ff.Envelope(0, 0, 0, 0)),
            direction=0xC000
        )

        self.effect_id = self.device.upload_effect(self.effect)
        self.device.write(ecodes.EV_FF, self.effect_id, 1)

        rospy.Subscriber(self.ns + "steer_feeback", SteerCtl, self.steer_feedback)
        rospy.Timer(rospy.Duration(DEVICE.LOOP_RATE), self.loop)
    
    def steer_feedback(self, msg):        
        if not self.p_des == msg.position or self.k_des == msg.springrate:                    
            self.p_des = msg.position
            self.k_des = abs(msg.springrate)                        
        
    def loop(self, _):
        p_raw = self.device.absinfo(self.axis_code)
        p_act = (2 * (p_raw.value - (self.axis_max + self.axis_min) * 0.5) / (self.axis_max - self.axis_min))

        self.p_err, t_corr, attack_length = self.getTorque(p_act, self.p_des, self.k_des, self.p_err)

        self.setTorque(t_corr, attack_length)

    def getTorque(self, p_des, p_act, k, p_err_old):
        p_err = p_des - p_act
        dp_err = p_err - p_err_old
        
        mul = 1 if p_err >= 0 else -1       

        t_corr = 0
        attack_length = 0

        if abs(p_err) > DEVICE.TOLERANCE:
            t_corr = mul * limit(k * abs(p_err), DEVICE.MIN_TORQUE, DEVICE.MAX_TORQUE)
            attack_length = DEVICE.LOOP_RATE
        
        return p_err, t_corr, attack_length

    def setTorque(self, t_corr, attack_length):
        torque_level = int(0x7FFF * t_corr)
        self.effect.u.constant.level = torque_level
        self.effect.u.constant.envelope.attack_length = int(attack_length * 1000)
        self.effect.u.constant.envelope.fade_length = int(attack_length * 1000)

        try:
            self.device.erase_effect(self.effect_id)
            self.effect_id = self.device.upload_effect(self.effect)
            self.device.write(ecodes.EV_FF, self.effect_id, 1)
        except Exception as e:
            rospy.logerr(f"Failed to upload force effect: {e}")

if __name__ == "__main__":
    try:
        sNode = SteeringFeedback()
        rospy.spin()

    except rospy.ROSInterruptException:
        pass

File: steerforce.cpp
#include <ros/ros.h>
#include <linux/input.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <fcntl.h>
#include <math.h>

#include "steering_data/SteerCtl.h"

class SteeringFeedback {

private:
    ros::Subscriber sub_target;
    ros::Timer timer;

    // rosparam
    std::string DEVICE_NAME = "/dev/input/event20";
    double DEVICE_TIME_STEP = 0.1;
    double DEVICE_TOLERANCE = 0.01;
    double DEVICE_MAX_TORQUE = 0.5;
    double DEVICE_MIN_TORQUE = 0.3;
    
    // device info
    int device_handle;
    int axis_code = ABS_X;
    int axis_min;
    int axis_max;

    double pAct;

    double pDes; // Desired Position
    double kDes; // Desired Spring Rate
    
    double pErr; 
    double tCorr; // Corrective Torque
    double atk_len;
    
    struct ff_effect ffEffect;

public:
    SteeringFeedback();
    ~SteeringFeedback();

private:
    void initDevice();
    void steerFeedback(const steering_data::SteerCtl &msg);
    void loop(const ros::TimerEvent&);
       
    void getTorque(const double &current_position);    
    void setTorque(const double &torque, const double &attack_length);
    
    // Helpers
    int testBit(int bit, unsigned char *array);
    double limit(const double &a,const double &a_min,const double &a_max);
};


SteeringFeedback::SteeringFeedback() {
    ros::NodeHandle ros_node;

    std::string ns = ros_node.getNamespace();
    
    ros_node.getParam("deviceName", DEVICE_NAME);
    ros_node.getParam("timeStep", DEVICE_TIME_STEP);
    ros_node.getParam("tolerance", DEVICE_TOLERANCE);

    ros_node.getParam("maxTorque", DEVICE_MAX_TORQUE);
    ros_node.getParam("minTorque", DEVICE_MIN_TORQUE);    
    
    pAct = 0;
    pDes = 0;
    kDes = 1;
    
    pErr = 0;
    tCorr = 0;
    atk_len = 0;

    initDevice();

    ros::Duration(1).sleep();

    timer = ros_node.createTimer(ros::Duration(DEVICE_TIME_STEP), &SteeringFeedback::loop, this);
    
    sub_target = ros_node.subscribe(ns + "steer_feeback", 1, &SteeringFeedback::steerFeedback, this);
}

SteeringFeedback::~SteeringFeedback() {
    ffEffect.type = FF_CONSTANT;
    ffEffect.id = -1;
    ffEffect.u.constant.level = 0;
    ffEffect.direction = 0;

    // upload ffEffect
    if (ioctl(device_handle, EVIOCSFF, &ffEffect) < 0) {
        std::cout << "Failed to upload ffEffect" << std::endl;
    }
}

// initialize force feedback device
void SteeringFeedback::initDevice() {
    // setup device
    unsigned char key_bits[1+KEY_MAX/8/sizeof(unsigned char)];
    unsigned char abs_bits[1+ABS_MAX/8/sizeof(unsigned char)];
    unsigned char ff_bits[1+FF_MAX/8/sizeof(unsigned char)];
    struct input_event event;
    struct input_absinfo abs_info;

    device_handle = open(DEVICE_NAME.c_str(), O_RDWR|O_NONBLOCK);
    if (device_handle < 0) {
        std::cout << "ERROR: cannot open device : "<< DEVICE_NAME << std::endl;
        exit(1);
    }
    
    memset(abs_bits, 0, sizeof(abs_bits));
    if (ioctl(device_handle, EVIOCGBIT(EV_ABS, sizeof(abs_bits)), abs_bits) < 0) {
        std::cout << "ERROR: cannot get abs bits" << std::endl;
        exit(1);
    }
    
    memset(ff_bits, 0, sizeof(ff_bits));
    if (ioctl(device_handle, EVIOCGBIT(EV_FF, sizeof(ff_bits)), ff_bits) < 0) {
        std::cout << "ERROR: cannot get ff bits" << std::endl;
        exit(1);
    }
    
    if (ioctl(device_handle, EVIOCGABS(axis_code), &abs_info) < 0) {
        std::cout << "ERROR: cannot get axis range" << std::endl;
        exit(1);
    }

    axis_max = abs_info.maximum;
    axis_min = abs_info.minimum;
    if (axis_min >= axis_max) {
        std::cout << "ERROR: axis range has bad value" << std::endl;
        exit(1);
    }

    if(!testBit(FF_CONSTANT, ff_bits)) {
        std::cout << "ERROR: force feedback is not supported" << std::endl;
        exit(1);

    }
    
    std::cout << "Device Ready" << std::endl;

    // Turn off the default auto centering.
    memset(&event, 0, sizeof(event));

    event.type = EV_FF;
    event.code = FF_AUTOCENTER;
    event.value = 0;

    if (write(device_handle, &event, sizeof(event)) != sizeof(event)) {
        std::cout << "Failed to disable autocentering" << std::endl;
        exit(1);
    }

    // init effect and get effect id
    memset(&ffEffect, 0, sizeof(ffEffect));

    ffEffect.type = FF_CONSTANT;
    ffEffect.id = -1; // initial value
    ffEffect.trigger.button = 0;
    ffEffect.trigger.interval = 0;
    ffEffect.replay.length = 0xffff;  // longest value
    ffEffect.replay.delay = 0; // delay from write(...)
    ffEffect.u.constant.level = 0;
    ffEffect.direction = 0xC000;
    ffEffect.u.constant.envelope.attack_length = 0;
    ffEffect.u.constant.envelope.attack_level = 0;
    ffEffect.u.constant.envelope.fade_length = 0;
    ffEffect.u.constant.envelope.fade_level = 0;

    if (ioctl(device_handle, EVIOCSFF, &ffEffect) < 0) {
        std::cout << "Failed to upload ffEffect" << std::endl;
        exit(1);
    }

    // start ffEffect
    memset(&event, 0, sizeof(event));

    event.type = EV_FF;
    event.code = ffEffect.id;
    event.value = 1;
    
    if (write(device_handle, &event, sizeof(event)) != sizeof(event)) {
        std::cout << "Failed to start event" << std::endl;
        exit(1);
    }
}

// get target information of wheel control from ros message
void SteeringFeedback::steerFeedback(const steering_data::SteerCtl &msg) {
    if (pDes != msg.position || kDes != fabs(msg.springrate)){
        pDes = limit(msg.position, -1.0, 1.0);
        kDes = fabs(msg.springrate);
    }
}

// update input event with timer callback
void SteeringFeedback::loop(const ros::TimerEvent&) {
    struct input_event event;
    
    // get current state
    while (read(device_handle, &event, sizeof(event)) == sizeof(event)) {
        if (event.type == EV_ABS && event.code == axis_code) {
            pAct = (event.value - (axis_max + axis_min) * 0.5) * 2 / (axis_max - axis_min);
        }
    }

    getTorque(pAct);    
    setTorque(tCorr, atk_len);    
}

void SteeringFeedback::getTorque(const double &pAct) {    
    double pErr = pDes - pAct;

    if (fabs(pErr) < DEVICE_TOLERANCE) {
        tCorr = 0.0;
        atk_len = 0.0;
    } else {
        double mul = (pErr >= 0.0) ? 1.0 : -1.0;

        tCorr =  mul * limit(fabs(kDes * pErr), DEVICE_MIN_TORQUE, DEVICE_MAX_TORQUE);
        atk_len = DEVICE_TIME_STEP;
    }

    std::cout<< pErr << ", " << tCorr << std::endl;
}

// update input event with writing information to the event file
void SteeringFeedback::setTorque(const double &torque, const double &attack_length) {
    // set effect
    ffEffect.u.constant.level = 0x7fff * torque;
    ffEffect.direction = 0xC000;    
    ffEffect.u.constant.envelope.attack_length = attack_length;    
    ffEffect.u.constant.envelope.fade_length = attack_length;
    
    // ffEffect.u.constant.envelope.fade_level = 0;
    // ffEffect.u.constant.envelope.attack_level = 0; /* 0x7fff * force / 2 */

    // upload effect
    if (ioctl(device_handle, EVIOCSFF, &ffEffect) < 0) {
        std::cout << "Failed to upload effect" << std::endl;
    }
}

// util for initDevice()
int SteeringFeedback::testBit(int bit, unsigned char *array) {
    return ((array[bit / (sizeof(unsigned char) * 8)] >> (bit % (sizeof(unsigned char) * 8))) & 1);
}

double SteeringFeedback::limit(const double &a, const double &a_min, const double &a_max){
    return(std::max(std::min(a, a_max), a_min));
}

int main(int argc, char **argv ){
    ros::init(argc, argv, "steering_feedback");
    SteeringFeedback sNode;
    ros::spin();
    return(0);
}